Feature/16 image carousel

fromITzerotoIThero opened the pull request:
Created nearly working ImageCarousel with the use of a random image array, yet without the image indicator. 
Arrows for changing the image work as expected with the use of a "hidden" class - a hint on how to make them appear in the right place within the container very much appreciated :)

piotrwicijowski commented:
On how to make the arrows appear in the right place
![image](https://github.com/fromITzerotoIThero/IBOS/assets/18663652/2ecb5855-67a0-487a-b602-d89082d75100)

fromITzerotoIThero commented:
So I've kind of completed the ImageCarousel and the ImageIndicator components, they work as expected, but I would appreciate your advice on how to make them "cleaner":

- do you think the logic is to be left that way or it would use some tweaks?
- how do you feel about the naming of props? Is it ok that way or should is it better practice to name props the same as their value if it's possilble?
- lengthy CSS selectors (I wasn't supposed to spend time mastering it so they're made simply to work) - how does a senior dev feel when looking at it - "icky" or "fine"?

We can discuss it live of course if you prefer not to write down all the answers :)

fromITzerotoIThero commented:
W., 

I've 2 questions:

1. I've updated variable and class names within ImageCarousel and pushed the changes, yet I'm not sure why the validation we did for `imageArray === null || imageArray.length === 0` breaks the page... I've added extra condition "imageArray === undefined" (because that's what the console was throwing as an error) but that wouldn't help either. What am I missing? 
2. In order to make a PR I should make final code that's gonna pull data from supabase, or should I leave it with the example Unsplash array - we've discussed it, but I forgot how it's meant to be...

piotrwicijowski commented:
1. Checking for `imageArray === undefined` helps. The reason why the old fix does not work in the current situation, is that since you do not pass the `imageArray` props, it turns out to be `undefined`, and the old checks did not cover that scenario. In such cases I tend to use `imageArray == null` (notice the double equals) - this check actually verifies two conditions - if it is null or if it is undefined. That's imho the only situation in which double equals `==` is useful
2. Nope, this PR only is for creating the component. However, extract the unsplash array out of the body of the component, and call the component using that array passed through the props.

fromITzerotoIThero commented:
Oh this is so neat! I checked for extra `imageArray === undefined` and it does work now, I must have done it somehow wrong, but I'm gonna go with `imageArray == null` :)

Ayay, extraction done!

piotrwicijowski commented on src/components/ImageCarousel/ImageCarousel.jsx:
```diff
@@ -6,47 +6,67 @@ import ChevronRightIcon from '@mui/icons-material/ChevronRight';
+
+    const isFirstImage = currentIndex === 0;
+    const isLastImage = currentIndex === imageArray.length - 1;
+    const leftArrow = isFirstImage ? styles.hidden : styles.arrow;
```
`leftArrow` what? Based on the rest of the code I would have to guess that this variable contains the class name of the left arrow, so it would be best if the name reflected that (same goes for `rightArrow`)

fromITzerotoIThero replied:
W., 
in situations when I'm not a 100% sure if the code changes I'm about to make based on your review, what's the preferred (from a code reviewer's perspective) approach:

1. Follow my intuition, commit changes and resolve conversation?
2. Ask in comments like this one and after your green light commit and resolve?

piotrwicijowski replied:
You can always ask. I may have not been clear on what I meant - I was confused by the variable name `leftArrow`, and I would recommend to change it to something more reflective of the contents of the variable, e.g. `leftArrowClassName`

fromITzerotoIThero replied:
OK, so I'll be asking in the PR then :) 
Thanks so so much for all the comments, I've just realized that refactoring the code based on your PR comments is just such a cool way of learning! I appreciate every single one!

piotrwicijowski commented on src/components/ImageCarousel/ImageCarousel.jsx:
```diff
@@ -6,47 +6,67 @@ import ChevronRightIcon from '@mui/icons-material/ChevronRight';
+    const leftArrow = isFirstImage ? styles.hidden : styles.arrow;
+    const rightArrow = isLastImage ? styles.hidden : styles.arrow;
+
+    const previousImage = () => {
```
If this is a function, then I would recommend to name it as a verb, e.g. `goToPreviousImage`. If it is just a noun, then there may be a wrong impression that the function would return the value (or url?) of the previous image, but instead it changes the current image.

piotrwicijowski commented on src/components/ImageCarousel/ImageCarousel.jsx:
```diff
@@ -6,47 +6,67 @@ import ChevronRightIcon from '@mui/icons-material/ChevronRight';
+        setCurrentIndex(newIndex);
+    };
+
+    const goToImage = (imageIndex) => {
```
Is this method even needed? It works exactly the same as `setCurrentIndex`, doesn't it?

fromITzerotoIThero replied:
Yep, it does! :)

piotrwicijowski commented on src/components/ImageCarousel/ImageCarousel.jsx:
```diff
@@ -6,47 +6,67 @@ import ChevronRightIcon from '@mui/icons-material/ChevronRight';
-            {imgArray != null
+            {imageArray != null
                 ?
                 <div>
```
Can we use a `<>` instead of `<div>` here? It would reduce the number of divs in divs in divs ;)

fromITzerotoIThero replied:
Oh that's a good idea! Is there a general rule you'd recommend when using `fragments`, like - 'whenever you don't need to set a class go with a fragment'?

piotrwicijowski commented on src/components/ImageCarousel/ImageCarousel.jsx:
```diff
@@ -6,47 +6,67 @@ import ChevronRightIcon from '@mui/icons-material/ChevronRight';
+                            <ChevronLeftIcon className={leftArrow} onClick={previousImage} />
+                            <ChevronRightIcon className={rightArrow} onClick={nextImage} />
+
+                            <div className={styles.indicator}>
```
It looks to me that the `arrow_container` div contains not only arrows but also the indicator. I'd suggest to have them as separate divs, so that the positioning and styling of one would not depend on the other (e.g. imagine a scenario in which at some point, if there are >20 images, we would need to hide the indicators, so we would not want the styling of the arrows to break because of that)

fromITzerotoIThero replied:
That sounds very future proof, I love it!

fromITzerotoIThero replied:
Oh, I have a problem here - I made the positioning of the Indicator work, yet through (say, not my favorite) trial&error method - the `bottom: 0.5rem` relates to what? Looks to me like the parent `<div className={styles.container}>` with 'position: relative`. 
For the `position: absolute` to work, we need to set `position: relative` on the parent, but would that work if `position: relative` were set on some ancestor (without any `position: relative` in between)?

piotrwicijowski replied:
Yes, `absolute` will work against the first explicitly positioned element that is in its hierarchy, so it does not matter if that's the elements parent, grandparent, or anything even higher, we do not need any `relative` in between.

fromITzerotoIThero replied:
I'm not sure I understand - so it's like I have `absolute` set on an element, it's "taken out" of the normal flow and it kind of like looks for the nearest ancestor that has any `relative` position set on it?

fromITzerotoIThero replied:
Oh, and 1 more thing!
At the very bottom of the PR there's this info about 1 more "Change requested", but I don't know what it refers to...
![pr](https://github.com/fromITzerotoIThero/IBOS/assets/109156809/c743c140-baf1-4b76-9fa4-5460e9f7c847)

piotrwicijowski commented on src/components/ImageIndicator/ImageIndicator.module.css:
```diff
@@ -1,3 +1,11 @@
-.icon {
-    font-size: 8px;
+.img_indicator > .dot {
```
Make sure to give the indicator a `cursor: pointer` rule so that it would be clear that it is clickable

piotrwicijowski commented on src/components/ImageIndicator/ImageIndicator.jsx:
```diff
@@ -1,14 +1,15 @@
     return (
-        <div>
-            <CircleIcon className={styles.icon}/>
+        <div className={styles.img_indicator}>
```
you can skip the `img_` prefix in the class name

fromITzerotoIThero commented:
W., 

Should I delete a branch once merged? I guess that's what we did last time, but I wanted to make sure :)`

piotrwicijowski commented:
Yes, it is a common pattern to delete a feature branch once the PR is merged.
